amethyst_input/src/input_handler.rs

```diff
@@ -204,6 +215,111 @@ where
         }
     }
 
+    /// Updates the input handler with a new controller event.
+    ///
+    /// Called internally from SdlEventsSystem when using sdl_controller feature.
+    /// You should invoke it in your system if you provide
+    /// your own controller input implementation.
+    pub fn send_controller_event(
+        &mut self,
+        event: &ControllerEvent,
+        event_handler: &mut EventChannel<InputEvent<AC>>,
+    ) {
+        use self::ControllerEvent::*;
+
+        match *event {
+            ControllerAxisMoved { which, axis, value } => {
+                if let Some(controller_id) = self.controller_idx_to_id(which) {
+                    self.controller_axes
+                        .iter_mut()
+                        .find(|(id, a, _)| *id == controller_id && *a == axis)
+                        .map(|entry| entry.2 = value)
+                        .unwrap_or_else(|| {
+                            self.controller_axes.push((controller_id, axis, value));
+                        });
+                    event_handler.single_write(event.into());
+                }
+            }
+            ControllerButtonPressed { which, button } => {
+                if let Some(controller_id) = self.controller_idx_to_id(which) {
+                    if self
+                        .pressed_controller_buttons
+                        .iter()
+                        .all(|&(id, b)| id != controller_id || b != button)
+                    {
+                        self.pressed_controller_buttons
+                            .push((controller_id, button));
+                        event_handler.iter_write(
+                            [
+                                event.into(),
+                                ButtonPressed(Button::Controller(controller_id, button)),
```

Comment:
Why 2 events that both represent a button press? 1 is the raw event and the second one is the mapped input using the configuration?

Reply:
I'm not sure really, I tried to mimic the current behaviour of other input methods, like mouse buttons. There are usually two separate events for each kind, one using `Button` and one using specific dedicated struct.

Reply:
@Xaeroxe is that how its supposed to be?

Reply:
@jojolepro Yeah it is.  You want to send both because you're not always certain what the user is looking for.

Reply:
Although come to think of it I don't think it's actually necessary, the API could just send the `Button` version of the event.